### PR TITLE
Introduce FakeClock and FakeSleeper

### PR DESCRIPTION
**Pull request by jart**
These classes are useful for testing many types of production code, without introducing nondeterminism or slowness.

### CODE REVIEW

File: tensorboard/test_util.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+  def __call__(self):
+    """Returns current fake time.
+
+    Returns:
+      A float of seconds since UNIX epoch in zulu time.
+
+    :rtype: float
+    """
+    with self._lock:
```
Comment (wchargin) [inline tensorboard/test_util.py]: Is this lock necessary? (I'm fine with keeping it even if it's not, but the only thing that I can imagine it protecting against is tearing, and this is something that Python should deal with.)

File: tensorboard/test_util.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  def __init__(self, clock):
+    """Creates new instance.
+
+    Args:
+      clock: A FakeClock instance.
+
+    :type clock: FakeClock
+    """
+    self._clock = clock
+    self._lock = threading.Lock()
```
Comment (wchargin) [inline tensorboard/test_util.py]: Not needed?

File: tensorboard/test_util.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+import threading
+
+
+class FakeClock(object):
+  """Fake version of time.time for testing clock code.
+
+  Classes that call time.time are encouraged to have a clock
+  constructor parameter, which can be swapped out with an instance of
+  this class for testing.
```
Comment (wchargin) [inline tensorboard/test_util.py]: And we can `tf.test.StubOutForTesting` those that don't!

Comment (jart) [inline tensorboard/test_util.py]: Indeed.

File: tensorboard/test_util.py (hunk 1 above)
Comment (jart) [inline tensorboard/test_util.py]: The honest answer is I don't know. I do know that Python doesn't have `volatile` or atomic numbers. I know it's possible that a write in one thread might not be visible to other threads afterwards if it's cached in L1 or L2. I don't know if this applies to Python, or if the GIL will keep us safe, even though some pythons don't have GIL. But what I do know is that the lock will keep us safe from whatever it is I don't know.

I'm also not 100% sure I know what you mean by tearing. I'm assuming that's related to why atomic numbers exist?

Comment (wchargin) [inline tensorboard/test_util.py]: That all makes sense, so I agree with your call to keep it.

By "tearing" I refer to the phenomenon by which a read of a 64-bit number can result in the lower 32 bits coming from one point in time and the higher 32 bits from another point in time.

As far as I know, the atomic primitives provided by Doug and the JSR-166 team are more targeted toward writes/CAS than reads—indeed, [the OpenJDK implementation of `AtomicInteger#get()`](http://www.grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/util/concurrent/atomic/AtomicInteger.java#AtomicInteger.get%28%29) does not lock. (As you pointed out, this doesn't mean that the lock isn't required in Python.)

**Pull request merged**